pages/blog/socketio-automatic-docs.md

```diff
@@ -0,0 +1,553 @@
+```python
+# instead of `from flask_socketio import SocketIO`
+from sio_asyncapi import AsyncAPISocketIO as SocketIO
+```
+3. Add additional argument to `AsyncAPISocketIO` constructor
+```python
+socketio = SocketIO(
+    app,
+    validate=True,
+    generate_docs=True,
+    version="1.0.0",
+    title="Tick-Tack-Toe API",
+    description="Tick-Tack-Toe Game API",
+    server_url="http://localhost:5000",
+    server_name="TICK_TACK_TOE_BACKEND",
+)
+```
+5. Tell `@socketio.on` decorator to get models from typehint.
+> Note: you can also pass `request_model` and `response_model` arguments to `@socketio.on` decorator instead of using typehints.
+```python
+@socketio.on('make_move', get_from_typehint=True)
+```
+Now type annotations will be used to generate AsyncAPI specification and validate incoming and outgoing messages. Note that the return value from a function is not data sent by `emit` function but rather `acknowledge` value that is received by the client.
+
+6. Add `on_emit` decorator to register/document a SocketIO emit event. Since we are not defining `emit` and only calling it as a function, we need to tell SIO-AsyncAPI what to expect when `emit` is called. E.g
+```python
+class GameCreatedData(BaseModel):
+    game_id: int
+
+@socketio.doc_emit('game_created', GameCreatedData)
+@socketio.on('create_game')
+def create_game():
+  ...
+    emit('game_created', {'game_id': game_id})
+```
+
+### Get AsyncAPI specification
+Now we can get AsyncAPI specification by calling `socketio.asyncapi_doc.get_yaml()` function. Here is how rendered specification looks like:
+
+![Figure 1:](/img/posts/socketio-automatic-docs/sio-asycnapi-pic1.webp)
+
+### Validation and Error handling
+SIO-AsyncAPI will automatically validate incoming and outgoing messages. If a message is not valid, it will raise one of these 3 exceptions: `EmitValidationError`, `RequestValidationError` or `ResponseValidationError`.
+
+Flask-SocketIO has a decorator for default error handling `@socketio.on_error_default` that we can use. E.g.:
+```python
+@socketio.on_error_default
+def default_error_handler(e: Exception):
+    """
+    Default error handler. It called if no other error handler defined.
+    Handles RequestValidationError, EmitValidationError and ResponseValidationError errors.
+    """
+    if isinstance(e, RequestValidationError):
+        logger.error(f"Request validation error: {e}")
+        return {'error': str(e)}
+    elif isinstance(e, ResponseValidationError):
+        logger.critical(f"Response validation error: {e}")
+        raise e
+    if isinstance(e, EmitValidationError):
+        logger.critical(f"Emit validation error: {e}")
+        raise e
+    else:
+        logger.critical(f"Unknown error: {e}")
+        raise e
+```
+Instead of re-raising exceptions, we can return some error which is interpreted as `acknowledge` value sent to the client. That's what we are doing in the example above when there is a RequestValidationError.
```

Comment:
```suggestion
Instead of re-raising exceptions, we can return some error interpreted as an `acknowledge` value sent to the client. That's what we do in the example above when there is a `RequestValidationError`.
```